src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -324,38 +188,42 @@ type MessageDbContext(client: MessageDbClient, batchOptions: BatchOptions) =
     member val BatchOptions = batchOptions
 
     member _.TokenEmpty = Token.create -1L
-    member _.LoadBatched(log, streamName, requireLeader, tryDecode, ct): Task<StreamToken * 'event[]> = task {
-        let! version, events = Read.loadForwardsFrom log client.ReadRetryPolicy client.Reader batchOptions.BatchSize batchOptions.MaxBatches streamName 0L requireLeader ct
-        return Token.create version, Array.chooseV tryDecode events }
-    member _.LoadLast(log, streamName, requireLeader, tryDecode, ct): Task<StreamToken * 'event[]> = task {
-        let! version, events = Read.loadLastEvent log client.ReadRetryPolicy client.Reader requireLeader streamName None ct
-        return Token.create version, Array.chooseV tryDecode events }
-    member _.LoadSnapshot(log, category, streamId, requireLeader, tryDecode, eventType, ct) = task {
+    member _.LoadBatched(streamName, requireLeader, tryDecode, fold, initial, ct): Task<struct(StreamToken * 'state)> = task {
+        let! version, state = Read.loadForwardsFrom fold initial tryDecode client.ReadRetryPolicy client.Reader batchOptions.BatchSize batchOptions.MaxBatches streamName 0L requireLeader ct
+        return struct(Token.create version, state) }
+    member _.LoadLast(streamName, requireLeader, tryDecode, fold, initial, ct): Task<struct(StreamToken * 'state)> = task {
+        let! version, events = Read.loadLastEvent client.ReadRetryPolicy client.Reader requireLeader streamName None ct
+        return struct(Token.create version, events |> Array.chooseV tryDecode |> fold initial) }
+    member _.LoadSnapshot(category, streamId, requireLeader, tryDecode, eventType, ct) = task {
         let snapshotStream = Snapshot.streamName category streamId
-        let! _, events = Read.loadLastEvent log client.ReadRetryPolicy client.Reader requireLeader snapshotStream (Some eventType) ct
-        return Snapshot.decode tryDecode events }
+        let! _, events = Read.loadLastEvent client.ReadRetryPolicy client.Reader requireLeader snapshotStream (Some eventType) ct
+        let decoded = Snapshot.decode tryDecode events
+        let act = Activity.Current
+        if act <> null then
+          let version = match decoded with ValueNone -> -1L | ValueSome (t, _) -> t.version
+          act.SetTag("eqx.snapshot_version", version) |> ignore
```

Comment:
better to arrive at a dictionary of names by moving out to tracing.fs ?

Reply:
Definitely. Wondering if we should expose the tag names as literals rather than extension methods?

Reply:
i like the extension methods - means you can hop across to see how diff stores implement things (though as I say that I realise it would be no different for literals), and it lets you maximise the amount of lunk not in your face if you're trying to absorb the flow.
For the tests, I think you can have a block of literals up the top of the test file, but it's not harmful for the code to be direct
Having said that, once you're running it, being able to grep for a tag name in the actual code becomes valuable.
Main thing is to pick one strategy at a time; then you can simply list off the pros for doing that when asked!